src/operators/tap.ts

```diff
@@ -74,49 +76,62 @@ class DoOperator<T> implements Operator<T, T> {
  * @ignore
  * @extends {Ignored}
  */
-class DoSubscriber<T> extends Subscriber<T> {
 
-  private safeSubscriber: Subscriber<T>;
+class TapSubscriber<T> extends Subscriber<T> {
+  private _context: any;
+
+  private _tapNext: ((value: T) => void) = noop;
+
+  private _tapError: ((err: any) => void) = noop;
+
+  private _tapComplete: (() => void) = noop;
 
   constructor(destination: Subscriber<T>,
-              nextOrObserver?: PartialObserver<T> | ((x: T) => void),
-              error?: (e: any) => void,
+              observerOrNext?: PartialObserver<T> | ((value: T) => void),
+              error?: (e?: any) => void,
               complete?: () => void) {
-    super(destination);
-
-    const safeSubscriber = new Subscriber<T>(nextOrObserver, error, complete);
-    safeSubscriber.syncErrorThrowable = true;
-    this.add(safeSubscriber);
-    this.safeSubscriber = safeSubscriber;
-  }
+      super(destination);
+      this._tapError = error || noop;
+      this._tapComplete = complete || noop;
+      this._tapNext = noop;
```

Comment:
83 already assigned into noop, do we need double init?

Reply:
Fixed.